tests/v1/test_stats.py

```diff
@@ -0,0 +1,281 @@
+        prompt="test_prompt",
+        mm_inputs=None,
+        mm_placeholders=None,
+        eos_token_id=None,
+        arrival_time=arrived_ts,
+        lora_request=None,
+    )
+    input_processed_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.INPUT_PROCESSED,
+        monotonic_ts_s=input_processed_ts,
+        engine_request=engine_request,
+    )
+    stats.update_from(input_processed_update)
+    assert stats.input_processor_end_ts_s == input_processed_ts
+    assert stats.engine_request == engine_request
+    assert stats.last_updated_ts_s == input_processed_ts
+    assert stats.num_prompt_tokens == 6
+    assert stats.sampling_params == sampling_params
+
+    assert stats.first_token_ts_s is None
+    assert stats.prefill_ts_s is None
+
+    # Test QUEUED
+    queued_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.QUEUED,
+        monotonic_ts_s=queued_ts,
+    )
+    stats.update_from(queued_update)
+    assert stats.queued_ts_s == queued_ts
+    assert stats.last_updated_ts_s == queued_ts
+
+    # Test RUNNING
+    running_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.RUNNING,
+        monotonic_ts_s=running_ts,
+        new_prefill=True,
+        num_computed_tokens=3,
+        num_cached_tokens=1,
+    )
+    stats.update_from(running_update)
+    assert stats.prefill_ts_s == running_ts
+    assert stats.num_computed_tokens == 3
+    assert stats.num_cached_tokens == 1
+    assert stats.queue_duration_s == running_ts - queued_ts
+
+    # Test RUNNING again shouldn't update prefill_ts_s
+    running_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.RUNNING,
+        monotonic_ts_s=running_2_ts,
+        new_prefill=False,
+        num_computed_tokens=6,
+        num_cached_tokens=0,
+    )
+    stats.update_from(running_update)
+    assert stats.prefill_ts_s == running_ts
+    assert stats.num_computed_tokens == 6
+    # num_cached_tokens is not updated
+    assert stats.num_cached_tokens == 1
+    assert stats.last_updated_ts_s == running_2_ts
+    # prefill_start_ts_s_lst should only contain the first running/resumed
+    # running prefill update.
+    assert stats.prefill_start_ts_s_lst == [
+        running_ts,
+    ]
+
+    # Test DECODED
+    decoded_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.DECODED,
+        monotonic_ts_s=decoded_ts,
+        num_new_tokens=1,
+        token_perf_ts_ns=decoded_ts * 1e9,
+    )
+    stats.update_from(decoded_update)
+    assert stats.last_updated_ts_s == decoded_ts
+    # Since arrival
+    assert stats.first_token_latency_s == decoded_ts - arrived_ts
+    assert stats.num_output_tokens == 1
+    # Since first scheduled
+    assert stats.prefill_latency_s == 2
+
+    # Test DETOKENIZED
+    detokenized_update = RequestStatsUpdate(
```

Comment:
I do not think that this `DETOKENIZED` event can fit into the structure that we have in V1. Since the detokenizer is not running in `EngineCore` and actually runs concurrently with  the `EngineCore` loop.

Reply:
How are you intending for these timestamps to be used?

Reply:
At the moment, I don't think we have any high-level metric using this. I think this will directly yields info on how long a detokenizer runs for each token(e.g. `detokenized ts - decoding ts`), which I am not too sure how long it could be - if it's insignificant enough, I think merging detokenizing into decoding step in metric tracking seems fine too. 

In short, It's more for logical completeness than any existing usecase, and potentially yielding info on the detokenizer.